rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp

```diff
@@ -254,6 +254,82 @@ Java_org_ros2_rcljava_node_NodeImpl_nativeCreateTimerHandle(
   return jtimer;
 }
 
+JNIEXPORT void JNICALL
+Java_org_ros2_rcljava_node_NodeImpl_nativeGetNodeNames(
+  JNIEnv * env, jclass, jlong handle, jobject jnode_names_info)
+{
+  rcl_node_t * node = reinterpret_cast<rcl_node_t *>(handle);
+  if (!node) {
+    rcljava_throw_exception(env, "java/lang/IllegalArgumentException", "node handle is NULL");
+    return;
+  }
+
+  jclass list_clazz = env->GetObjectClass(jnode_names_info);
+  jmethodID list_add_mid = env->GetMethodID(list_clazz, "add", "(Ljava/lang/Object;)Z");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jclass node_info_clazz = env->FindClass("org/ros2/rcljava/graph/NodeNameInfo");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jmethodID node_info_init_mid = env->GetMethodID(node_info_clazz, "<init>", "()V");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID name_fid = env->GetFieldID(node_info_clazz, "name", "Ljava/lang/String;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID namespace_fid = env->GetFieldID(node_info_clazz, "namespace", "Ljava/lang/String;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jfieldID enclave_fid = env->GetFieldID(node_info_clazz, "enclave", "Ljava/lang/String;");
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcutils_string_array_t node_names = rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t node_namespaces = rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t enclaves = rcutils_get_zero_initialized_string_array();
+
+  rcl_ret_t ret = rcl_get_node_names_with_enclaves(
+    node,
+    allocator,
+    &node_names,
+    &node_namespaces,
+    &enclaves);
+  RCLJAVA_COMMON_THROW_FROM_RCL(env, ret, "rcl_get_node_names_with_enclaves failed");
+
+  if (node_names.size != node_namespaces.size || node_names.size != enclaves.size) {
+    rcljava_throw_exception(
+      env,
+      "java/lang/IllegalStateException",
+      "names, namespaces and enclaves array leghts don't match");
+    goto cleanup;
```

Comment:
Now that we have the dependency on `rcpputils`, I guess we can used a scoped exit for cleanup.

Reply:
Done in https://github.com/osrf/ros2_java/pull/31/commits/5af448a6c06d3282acaaf24b1fe8b2a38fe924a8.